llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -7407,3 +7407,23 @@ bool CombinerHelper::matchZextOfTrunc(const MachineOperand &MO,
 
   return false;
 }
+
+bool CombinerHelper::matchNonNegZext(const MachineOperand &MO,
+                                     BuildFnTy &MatchInfo) {
+  GZext *Zext = cast<GZext>(MRI.getVRegDef(MO.getReg()));
+
+  Register Dst = Zext->getReg(0);
+  Register Src = Zext->getSrcReg();
+
+  LLT DstTy = MRI.getType(Dst);
+  LLT SrcTy = MRI.getType(Src);
+  const auto &TLI = getTargetLowering();
+
+  // Convert zext nneg to sext if sext is the preferred form for the target.
+  if (TLI.isSExtCheaperThanZExt(getMVTForLLT(SrcTy), getMVTForLLT(DstTy))) {
+    MatchInfo = [=](MachineIRBuilder &B) { B.buildSExt(Dst, Src); };
```

Comment:
```
if (!isLegalOrBeforeLegalizer({TargetOpcode::G_SEXT, {DstTy, SrcTy}}))
  return false;
```
We have to check whether the sext is legal or we are before the legalizer..

Reply:
It is only cheaper. You had to use MVT types to ask for cheapness. It is a DAGish call. You still have to query the GlobalIsel legalizer for legalness.